File: solutions/nim/clock/1/clock.nim
# Dedicated to Shree DR.MDD
import strformat, math

type
  Clock* = object
    hour*: range[0..23]
    minute*: range[0..59]
  Minutes* = distinct int

proc initClock*(h: int, m: int): Clock =
  let totalMins = (h * 60 + m).euclMod(24 * 60)
  return Clock(hour: totalMins div 60, minute: totalMins mod 60)

proc `$`*(clk: Clock): string =
  result = fmt"{clk.hour:02}:{clk.minute:02}"

proc `+`*(clk: Clock, mins: Minutes): Clock =
  return initClock(clk.hour, clk.minute + mins.int)

proc `-`*(clk: Clock, mins: Minutes): Clock =
  return initClock(clk.hour, clk.minute - mins.int)

File: solutions/go/clock/1/clock.go
// Shree DR.MDD
package clock
import "fmt"
const (
	hourToMin = 60
	dayToMin  = 24 * hourToMin
)
type Clock struct {
	totalMin int
}
func New(h, m int) Clock {
	val := (h*hourToMin + m) % dayToMin
	if val < 0 {
		val += dayToMin
	}
	return Clock{val}
}
func (c Clock) Add(m int) Clock {
	return New(0, c.totalMin+m)
}
func (c Clock) Subtract(m int) Clock {
	return New(0, c.totalMin-m)
}
func (c Clock) String() string {
	return fmt.Sprintf("%02d:%02d", c.totalMin/hourToMin, c.totalMin%hourToMin)
}
